### ISSUE TITLE
`make -C stdlib depend` is broken

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/ocaml#11588]
it says it has nothing to do when it has. Work around:

```
rm stdlib/.depend
make -C stdlib depend
```

**Comment by dra27** [comment]
The fix for this was part of a PR which got closed, but which I’ve got a TODO item to transfer to a fresh PR. I expect to do that well before the Stale bot comments on this issue 😉

**Comment by gasche** [comment]
I am learning just now that this issue is probably about what has been making my life miserable for the last week or so.

(I have the diffuse impression that building the compiler with `make` is more and more broken as time passes, but I rarely bother stopping whatever I am actually trying to do to debug the problem, it is easier to just `make partialclean` and rebuild. (For example I lost count of the number of digest mismatches in compilerlibs/*.cma that have made my life slightly painful and have been obliterated this way). But *this* issue is evil because this plug-it-off-and-on-again approach does not actually seem to work.)

**Comment by shindere** [comment]
See anagram #11885 for a fix.

**Comment by shindere** [comment]
Gabriel Scherer (2023/01/11 07:54 -0800):
> (I have the diffuse impression that building the compiler with `make`
> is more and more broken as time passes, but I rarely bother stopping
> whatever I am actually trying to do to debug the problem it is easier
> to just `make partialclean` and rebuild.

Next time you encounter such a problem, I'd be happy to investigate if
you can provide a repro case. I don't think the recent changes on the build
system have made the situation worse than it was or, if that happened,
my belief is that it has been fixed shortly after it had been reported.
Of course it's a bit more difficult to fix problems that have not been
reported.

> (For example I lost count of
> the number of digest mismatches in compilerlibs/*.cma have made my
> life slightly painful and have been obliterated this way).

I think many of them, if not all of them, are due to the fact that
compilerlibs needs to be copied into `otherlibs/dynlink`. It's something
which is not very easy to solve but I am working on that so let's hope
no digest issue will remain afterwards. Or at least this should get rid
of some of them and it will be interesting to see which issues remain.

> But *this*
> issue is evil because this plug-it-off-and-on-again approach does not
> actually seem to work.)

Yeah. But see, reported and fix proposed just a few hours later so
things are not that hopeless. Sorry the issue has not been fixed
earlier. I apologize for that.

**Comment by gasche** [comment]
Thanks for the fix! (Nitpick: apparently the issue was first reported by @dbuenzli on September 30th 2022.)

I'm not trying to blame you (or @dra27) for the current state of things, but reporting these issues is difficult for several reasons:

- It's hard for me to know whether I did something "wrong" or the outcome should not actually be expected. For example, I semi-regularly `git checkout` from a branch to another that sits on an older trunk, with a configuration change in the meantime; but most of the time there was no config change so I don't distclean by default. But then `make` breaks, and I assume that it's my fault. Sometimes it is less clear, I'm not sure whether it is my fault or not.
- Some of those dependency bugs I sort of assume are well-known and just hard to fix -- we just mentioned compilerlibs which creates such issue. So if I observe something that I believe is a bug, I need to build some confidence that it is not an already-known bug.
- Reproducing these things is tricky, because often they depend on the intermediate state of my build. (How would I send a repro case? I would have to try to find a manual modification I can perform from just the sources to get in a similar failing state again. Hard and time-consuming.) 
- As I mentioned, when I see those bugs I am always in the middle of doing something else than I want to do, so I don't want to stop to report bugs, especially when there is probably an easy workaround (clean and rebuild). But then the workaround destroys the state in which the bug occurred in the first place.

In summary, it takes a lot of discipline to actually try to report this kind of bugs, and unfortunately I generally lack this discipline.

As you are asking for reports: I *think* there is an issue with the fact that, running `ocamltest` on a given test after a `make -j`, I often see bugs due to the fact that `tools/expect_cmo` artifacts are stale. I generally have to `make -C testsuite clean && make -C testsuite/tools expect_test.cmo` manually before I can actually run `ocamltest`. I wish that `make -j` would get me to a state where I can run tests without cleaning and rebuilding stuff. Should I report an issue about this?

**Comment by gasche** [comment]
I wonder if it would be possible to more thoroughly test our Makefile through an automated process. For example we could script the following: do a fuild build, then change a .mli file in the stdlib in a way that is innocuous but changes its digest, then do a parallel incremental build. (And then we could do the same for a .ml file, and we for each compiler-libs archive we could do the same for one of its files, and we could it for a file in each of the `otherlibs` system, etc.)

**Comment by shindere** [comment]
Gabriel Scherer (2023/01/11 12:10 -0800):
> Thanks for the fix! (Nitpick: apparently the issue was first reported
> by @dbuenzli on September 30th 2022.)

Yes, I noticed. Since @dra27 wrote that he had a fix I stopped
monitoring the issue and that's why I am apologizing. Not to say it's
David's fault. There were many many things to do re:multicore and OCaml
5.0 and David was way more involved than me in these things, that's why
I should have been more vigilant here.

> I'm not trying to blame you (or @dra27)

No problem, I believe you. Just that it felt so, because you wrote ``I
have the diffuse impression that building the compiler with `make` is
more and more broken``, which I read as meaning that there was a
regression. I find this unfair, especially if it's not backed by facts.

> for the current state of things, but reporting these issues is difficult for several reasons:
> 
> - It's hard for me to know whether I did something "wrong" or the
> outcome should not actually be expected (for example, I
> semi-regularly `git checkout` from a branch to another that sits on
> an older trunk, with a configuration change in the meantime; but
> most of the time there was no config change so I don't distclean by
> default.

It may not be safe to do that, indeed, and perhaps this is indeed not
supposed to work.

> But then `make` breaks, and I assume that it's my fault.)
> Sometimes it is less clear.

Have you considered using different worktrees? I assume you are aware of
the `git worktree` possibility?

I find it really convenient because it really makes you available to
work on several worktrees in parallel and in a safe way.

> - Some of those dependency bugs I sort of assume are well-known and
> just hard to fix -- we just mentioned compilerlibs which creates
> such issue. So if I observe something that I believe is a bug, I
> need to build some confident that that one is not a known bug.

They may be known but not well-known. In any case, to me such bugs
simply need to be fixed, once and for all and in a clean way so I realy
would like to hear about them. So please, assume I personnally don't
know about them. Most of the time you will be correct.

Should we somehow devote some time on such discussions? I do realise
they do not have a direct impact on users, but they still have because
these things make our lives as developers more difficult and thus make
us less efficient.

> - Reproducing these things is tricky, because often they depend on the
> - intermediate state of my build. (How would I send a repro case? I
> - would have to try to find a manual modification I can perform from
> - just the sources to get in a similar failing state again. Hard and
> - time-consuming.)

Yes, I agree. Would it work to just upload the full tree as it is
somewhere? I realise it would not work 100% of the time but perhaps it's
worth giving it a try?

> - As I mentioned, when I see those bugs I am always in the middle of
> doing something else than I want to do, so I don't want to stop to
> report bugs, especially when there is probably an easy workaround
> (clean and rebuild). But then the workaround destroys the state in
> which the bug occurred in the first place.

Yes yes, I understand. It should be straightforward to report. Shall we
try what I have suggested, i.e. just do a tar and uploadit somewhere and
then send a link to me?

> All this being said, it takes a lot of discipline to actually try to
> report this kind of bugs, and unfortunately I generally lack this
> discipline.

I understand. See whether the proposal above can help.

> As you are asking for reports: I *think* there is an issue with the
> fact that, running `ocamltest` on a given test after a `make -j`, I
> often see bugs due to the fact that `tools/expect_cmo` artifacts are
> stale.

What do you mean by stale?

> I generally have to `make -C testsuite clean && make -C
> testsuite/tools expect_test.cmo` manually before I can actually run
> `ocamltest`. I wish that `make -j` would get me to a state where I can
> run tests without cleaning and rebuilding stuff. Should I report an
> issue about this?

You could if you want but I don't think it's necessary.

My answer is that the problem you are having has basically to do with
dependencies between objects that live in different directories.
Basically, if ocamltest is enabled, then that should imply that the
tools under `testsuite/tools` also get build by `make -j`. I did try to
fix this but I felt that with the current state of the makefiles it was
hard, whereas it will become trivial after they all get merged, which
will happen in a few months I hope, let's say before the summer. Canyou
wait till then?

It's a not-so-easy question because, as I said, in the current situation
it's not that easy to fix. And, the time I will spend doing (or
reviewing) fixes I won't spend on the merge which should be the
long-term solution.

**Comment by gasche** [comment]
> you wrote ``I
have the diffuse impression that building the compiler with `make` is
more and more broken``, which I read as meaning that there was a
regression. I find this unfair, especially if it's not backed by facts.

Fact: dependency computation between stdlib modules was broken in March 2022, it was reported by an external contributor at the very end of September, and you are now fixing this in January 2023. Between March 2022 and January 2023, people changing dependencies between stdlib modules probably had a bad experience.

Again, I am not blaming anyone in particular for this situation, but I *do* have the diffuse impression that building the compiler is erroring in weird ways more often than it was, say, five years ago. I am not a build-system expert, but I think that `make` as a build tool does not actually scale to design build systems of the complexity we have here with the compatibility and configurability constraints we have set on ourselves, and the result is a complex mess that is full of bugs. (Or maybe there is a way to architect correct build systems with Make, but it's not the way our current build system is currently built.)

We regularly add complexity to the build system because we want to do more things with it (I'm certainly among the people that took part in this complexification), but this also introduces bugs. It looks like this is precisely what happened with this `.INTERMEDIATE` feature that was introduced in March 2022.

I haven't been following closely what @dra27 and yourself @shindere have been doing on the build system, but from a distance it looks like it goes in the general direction of making it more robust and less buggy. This is great, and I guess we have to accept to pay the price of occasional bugs introduced by this gradual change.

**Comment by shindere** [comment]
Gabriel Scherer (2023/01/11 13:26 -0800):
> > you wrote ``I
> have the diffuse impression that building the compiler with `make` is
> more and more broken``, which I read as meaning that there was a
> regression. I find this unfair, especially if it's not backed by facts.
> 
> Fact: dependency computation between stdlib modules was broken in
> March 2022, and that it was reported by an external contributor at the
> very end of September, and that you are now fixing this in January
> 2023. Between March 2022 and January 2023, people changing
> dependencies between stdlib modules probably had a bad experience.

That's one fact, yes. Just one.

> Again, I am not blaming anyone in particular for this situation, but I
> *do* have the diffuse impression that building the compiler is
> erroring in weird ways more often than it was, say, five years ago.

Perhaps we also have more demands. For instance, were parallel builds as
important five years ago than they are nowadays? I did'nt check but I'd
guess that, five years ago, FLambda was not merged, or was it? I am
saying this because FLambda is systematically compiled, whether used or
not, and this slows down the build and makes it even more important ot
have parallel builds working to compensate. I am also guessing that,
five years ago, the copying business of compilerlibs into dynlink was
not done so the build was more straightforward, but then there were
downsides for users, which is exactly what led to copying compilerlibs
around.

Also, I myself do not usually work with older versions of OCaml, like
@dra27 does when he backports features. What he reports, if I understand
him correctly, is that when you backport something you see how more
difficult it was in the past to do some stuff than it is now. Remember
the pre-configure era or, the time when there were two similar but
distinct build systems for Unix and Windows. You @gasche as a Unix user
may not have been impacted that much, but I am pretty sure the Windows
users were, because the build system they were using was not necessarily
up-to-date with hte more used Unix one. Also, thos having to maintain
the build system suffered.

> I am not a build-system expert, but I think that `make` as a build
> tool does not actually scale to design build systems of the complexity
> we have here with the compatibility and configurability constraints we
> have set on ourselves, and the result is a complex mess that is full
> of bugs. (Or maybe there is a way to architect correct build systems
> with Make, but it's not the way our current build system is currently
> built.)

As you may have guessed I believe more in your second hypothesis than in
the first one. Not to say that `make` is perfect, it definitely has its
pitfalls. But fur sure, we do not help ourselves by how we currenly use
it.

> We regularly add complexity to the build system because we want to do
> more things with it (I'm certainly among the people that took part in
> this complexification), but this also introduces bugs. It looks like
> this is precisely what happened with this `.INTERMEDIATE` feature that
> was introduced in March 2022.

I pledge guilty for having introduced it, and for not having payed
enough attention to this issue.

> I haven't been following closely what @dra27 and yourself @shindere
> have been doing on the build system, but from a distance it looks like
> it goes in the general direction of making it more robust and less
> buggy. This is great, and I guess we have to accept to pay the price
> of occasional bugs introduced by this gradual change.

Probably, yes. Again, currently the count of such bugs has just been
incremented from zero to one.

It also feels to me that the price we are paying, to quote your words,
is not only the price for the changes we are doing but also the price for
years and years of accumulated technical debt, of nobody actually being
in charge of this part of the system and of people rushing modifications
to achieve whath they needed but without being able to do it in a
maintainable and sustainable way. I am not blaming anybody myself
either, I am fully convinced people did the bet htey could. I am just
trying to say that , although you do not blame anybody, you seem to
blame the changes themselves, but these changes would not have been
necessary if, in the first place, the systme would have been in good
shape. That's why I find it unfair to blame the changes, even if you
don't blame the persons. Also because it's an area of the compiler not
many people want to touch, otherwise it would have been more maintained,
so franlky blames, even on changes, are not so nice to read.

**Comment by gasche** [comment]
> so franlky blames, even on changes, are not so nice to read.

Again, I don't think I blame anyone. I say that, for *my* experience, the build system is less reliable. I used appropriate words to express subjectivity and uncertainty ("I have the diffuse impression that...") and I say that *the build system* is not so reliable. I am blaming *the build system*.

When there are problems, I think that we should allow ourselves to mention them. If you feel guilty when I say that the build system is not so reliable, I am sorry about it and it is certainly not my intention, but I don't think that this should prevent me from mentioning the problem.

I encounter build failures when running "make" on a semi-regular basis, it is annoying, and I have the diffuse impression that it is more frequent than it was, say, five years ago. Fact.

The rest is interpretation. You could decide to  take this as an extra data point in support of your makefile refactoring work, or as a data point in favor of rewriting some of the messy make+sed+bash soup into a scripting language, or as a data point to support people trying to redo the build system using other tools, or as secretly blaming people even though I am explicitly saying I am not. Your choice.

**Comment by dra27** [comment]
I think it may be time to finish this discussion over a (virtual) coffee!

If my understanding of the original problem in https://github.com/ocaml/ocaml/pull/11421#discussion_r918008508 was correct, the issue affected the changing of dependencies _between_ stdlib modules only, not between the stdlib and the rest of the compiler (in particular compiler-libs). The dependencies between stdlib modules is something which doesn't usually change that often, which is why I hadn't particularly prioritised this issue (yet).

**Comment by shindere** [comment]
Gabriel Scherer (2023/01/11 12:40 -0800):
> I wonder if it would be possible to more thoroughly test our Makefile
> through an automated process. For example we could script the
> following: do a fuild build, then change a .mli file in the stdlib in
> a way that is innocuous but changes its digest, then do a parallel
> incremental build. (And then we could do the same for a .ml file, and
> we for each compiler-libs archive we could do the same for one of its
> files, and we could it for a file in each of the `otherlibs` system,
> etc.)

It's certainly possible to do at least a bit of CI on the build system
itself. Your suggestion feels already very sophisticated to me, but
that's not to say we shouldn't do it. I'd rather say that there are more
fundamental things we could test even before going that foar. For
instance, one possible first step could be to test that the build log
does not change in a way that goes unnoticed. So compare the build log
to an expected build log and see whether they differ (we introduced or
lost a flag unintendedly...).

**Comment by nojb** [comment]
This has been fixed in #11885
